### ISSUE TITLE
Bot should use kaids instead of usernames

### ISSUE THREAD

**Issue by MatthiasPortzel** [BlazeProgramming/Tucker#11]
Title says it all. Tucker should probably keep track of users' `kaid`s, which never change. Then he (it?) could look up information about the user with the Khan Academy api.

**Comment by 1kq2lo40lozpmb63** [comment]
Wow. That's genius. Love it. There are only two problems:
1) From my experience there seem to be 2 "types" of KAID used in the system (one starts with `kaid_`)
2) How would you tell users to get their KAIDs to input? It seems like a very tedious process, especially for the kids and newbies we have on the server.

**Comment by MatthiasPortzel** [comment]
I've worked with the KA api quite often. Technically speaking, everyone only has one kaid, and it starts with `kaid_`, so my kaid is `kaid_757721856896775939251306`. And the kaid would be looked up by the bot; the user would enter his/her username just like now.

**Comment by 1kq2lo40lozpmb63** [comment]
Good idea. Supported

**Comment by CosignCosine** [comment]
I agree, and that is how I was initially implementing the program. Originally, the !addid program got the username, sent it to the API, got the KAID, and added it. That was a LOT of code (and frankly it didn't work). I'm going to implement this when I have time. :)

**Comment by lukekrikorian** [comment]
So, I've been looking into this. [Khan Academy API Explorer](https://api-explorer.khanacademy.org/) this website is pretty nice. But, to get any info on users you need them to log in. Not sure how you would do that with a chat bot. Anyways, blaze's idea won't work.

**Comment by lukekrikorian** [comment]
I **have** been looking in to other ways to do this. There is [this](https://github.com/weo-edu/khan) Javascript library.

**Comment by 1kq2lo40lozpmb63** [comment]
You know you could just send a request to
/api/internal/profile?user=username or something like that. The kaid is
included in the response.

**Comment by 1kq2lo40lozpmb63** [comment]
Just checked; it's https://www.khanacademy.org/api/internal/user/profile?username=USERNAMEHERE

**Comment by lukekrikorian** [comment]
Oh wow. That's awesome.

**Comment by 1kq2lo40lozpmb63** [comment]
Yeah. It's a hidden API we all independently found a year back or something. Guardians say it's completely OK to use so it's even better :)

**Comment by 1kq2lo40lozpmb63** [comment]
Oh, and you can also pass in https://www.khanacademy.org/api/internal/user/profile?kaid=kaid_somethinghere to get data given the KAID.

**Comment by lukekrikorian** [comment]
An vice versa. That's awesome. But how would you access the info on the 'website' in a program?

**Comment by 1kq2lo40lozpmb63** [comment]
What do you mean?

**Comment by lukekrikorian** [comment]
Like, how would the program access the info at https://www.khanacademy.org/api/internal/user/profile?kaid=kaid_somethinghere

**Comment by 1kq2lo40lozpmb63** [comment]
Here's an example program: https://www.khanacademy.org/computer-programming/test/5571887270592512

**Comment by lukekrikorian** [comment]
Ok, cool. Is that possible with this bot?

**Comment by 1kq2lo40lozpmb63** [comment]
Of course, you can send a normal HTTP GET request through the already `require`d `request.js` library.

**Comment by lukekrikorian** [comment]
OK. Should I try and implement this?

**Comment by 1kq2lo40lozpmb63** [comment]
Sure, go ahead. I myself am too horrified by the messiness of the code that I don't think I'll edit it.

**Comment by lukekrikorian** [comment]
Ok 😃 . Can someone add me as a contributor?

**Comment by 1kq2lo40lozpmb63** [comment]
No, you can just make a pull request.

**Comment by lukekrikorian** [comment]
OK.

**Comment by CosignCosine** [comment]
I can't make everyone direct contributors because I would like the bot to not be vandalized, but you can certainly make a direct pull request ;)

**Comment by lukekrikorian** [comment]
How exactly do we want this to be implemented? If a new user join the bot automatically asks them what their KA username is?

**Comment by 1kq2lo40lozpmb63** [comment]
No, we'll use it like it is now: The user can enter a command with their username and it'll store it in the database.

**Comment by lukekrikorian** [comment]
OK, that makes sense.

**Comment by lukekrikorian** [comment]
(Newb Here)
What is endSub?

**Comment by 1kq2lo40lozpmb63** [comment]
It's a (TOTALLY UNNECESSARY) function that gets the rest of the message after `n` characters.

**Comment by lukekrikorian** [comment]
How would I access that in a variable?

**Comment by 1kq2lo40lozpmb63** [comment]
Umm...

``` javascript
var something1 = endSub(something2, something3);
```

**Comment by kingcodefish** [comment]
This is just one example of the abhorrent and horrendous code that Blaze has carefully crafted to annoy us to the end of days.

Also, another reason why Blaze should really document this stuff.

**Comment by 1kq2lo40lozpmb63** [comment]
Precisely proffered past participles present parried pounces pointed past present participant @ProgrammingBlaze.

**Comment by lukekrikorian** [comment]
Can you formulate the request.js call that outputs the kaid in a variable @raumaankidwai?

**Comment by 1kq2lo40lozpmb63** [comment]
Sure:

``` javascript
request("https://www.khanacademy.org/api/internal/user/profile?username=" + username, function (e, r, b) {
    if (!e && r.statusCode == 200) {
        kaid = JSON.parse(b).kaid;
    } else {
        // do something if there's an error
    }
});
```

**Comment by lukekrikorian** [comment]
Thanks!

**Comment by lukekrikorian** [comment]
Thanks so much!

*Luke Krikorian *

**Comment by 1kq2lo40lozpmb63** [comment]
No prob.

**Comment by lukekrikorian** [comment]
You meant to add a `var` before line 3, right?

**Comment by 1kq2lo40lozpmb63** [comment]
No, if you add a `var` the variable can't be accessed from outside the function. I meant it to be used like this:

``` javascript
var username = "something";
var kaid;
request("https://www.khanacademy.org/api/internal/user/profile?username=" + username, function (e, r, b) {
    if (!e && r.statusCode == 200) {
        kaid = JSON.parse(b).kaid;
    } else {
        // do something if there's an error
    }
});
// Now the variable `kaid` will have the kaid
```

**Comment by lukekrikorian** [comment]
Does [this](https://github.com/lukekrikorian/Tucker/blob/master/bot.js) look good?

**Comment by lukekrikorian** [comment]
Is anyone there?

**Comment by CosignCosine** [comment]
@raumaankidwai Why not create nested design within each callback? That way, if it is an asynchronous request, the variables will all be defined by the time the other commands are run.

request -> fs -> return

**Comment by lukekrikorian** [comment]
@BlazeProgramming did you get my pull request?

**Comment by CosignCosine** [comment]
@lukekrikorian That actually doesn't work. We want to store the kaid, not the username, remember? ;)

**Comment by lukekrikorian** [comment]
That's what the command does. ❓

**Comment by MatthiasPortzel** [comment]
@lukekrikorian A little more background: One of the most frustrating things about KA, in my opinion is the lack of documentation for the internal API. I created [this program](https://www.khanacademy.org/cs/i/6389992281473024) (with some help) to provide some documentation, but it's far from complete and has no UI. [Here's](https://github.com/BlazeProgramming/Tucker/blob/326cbec5b60c7f2f27a6f2ba841ae69203dd3298/bot.js#L822) an example from another location in this project where `request` and the internal API are used for the `!userdata` Tucker command. I would recommend using this as a template. Also, we would want to override the `!addid` [command](https://github.com/BlazeProgramming/Tucker/blob/326cbec5b60c7f2f27a6f2ba841ae69203dd3298/bot.js#L723). Also, we would need to change the `!kaname` command to get the username with a similar API call, rather than returning the kaid.
